src/DmaRequestCore.bsv

```diff
@@ -0,0 +1,99 @@
+import FIFOF::*;
+import PcieTypes::*;
+import DmaTypes::*;
+
+typedef 4096 BUS_BOUNDARY
+typedef 12   BUS_BOUNDARY_WIDTH
+
+typedef struct {
+    DmaRequestFrame dmaRequest;
+    Maybe#(DmaMemAddr) firstChunkLen;
+} ChunkRequestFrame deriving(Bits, Eq);                     
+
+Interface ChunkCompute;
+    FifoIn#(DmaRequestFrame) dmaRequests;
+    FifoOut#(DmaRequestFrame) chunkRequests;
+endinterface 
+
+module mkChunkComputer(ChunkCompute ifc);
+    FIFOF#(DmaRequestFrame) inputFifo <- mkFIFOF;
+    FIFOF#(DmaRequestFrame) outputFifo <- mkFIFOF;
+    FIFOF#(ChunkRequestFrame) splitFifo <- mkFIFOF;
+
+    Reg#(DmaMemAddr) newChunkPtrReg <- mkReg(0);
+    Reg#(DmaMemAddr) totalLenRemainReg <- mkReg(0);
+    Reg#(Bool) isSplittigReg <- mkReg(False);
+
+    function Bool hasBoundary(DmaRequestFrame request);
+        let highIdx = (request.startAddr + request.length) >> BUS_BOUNDARY_WIDTH;
+        let lowIdx = request.startAddr >> BUS_BOUNDARY_WIDTH;
+        return (highIdx > lowIdx);
+    endfunction
+
+    function DmaMemAddr getOffset(DmaRequestFrame request);
+        DmaMemAddr offset = zeroExtend(fromInteger(valueOf(BUS_BOUNDARY)) - pack(request.startAddr[BUS_BOUNDARY_WIDTH-1:0]));
+        return offset;
+    endfunction
+
+    rule getfirstChunkLen if(inputFifo.notEmpty && splitFifo.notFull);
+        let request = inputFifo.first;
+        inputFifo.deq;
+        let offset = getOffSet(request);
+        // firstChunkLen = offset % PCIE_TLP_BYTES
+        DmaMemAddr firstLen = zeroExtend(offset[valueOf(PCIE_TLP_BYTES_WIDTH)-1:0]);
```

Comment:
you have to read the PG213 to find out how to get the Max Payload Size(MPS) and Max Read Request Size(MRRS), those value is not const so you should do some refacrtor.

Reply:
MPS&MRRS are passed in the _Configuration Status Interface Port_ . They should be parameters of my module.
![image](https://github.com/BIGWJZ/blue-dmac/assets/49031111/4307d2ed-4b95-4039-87ff-69185b1509fa)

![image](https://github.com/BIGWJZ/blue-dmac/assets/49031111/0c78b4df-cb0d-45c0-98c5-58c0ccb7f133)